tests/lib/rules/prefer-expect-query-by.js

```diff
@@ -0,0 +1,37 @@
+'use strict';
+
+const RuleTester = require('eslint').RuleTester;
+const rule = require('../../../lib/rules/prefer-expect-query-by');
+
+const ruleTester = new RuleTester({
+  parserOptions: { ecmaVersion: 2015, sourceType: 'module' },
+});
+
+ruleTester.run('prefer-expect-query-by', rule, {
+  valid: [
+    { code: "expect(queryByText('Hello')).not.toBeInTheDocument()" },
+    { code: "expect(rendered.queryByText('Hello')).not.toBeInTheDocument()" },
+    { code: "expect(queryAllByText('Hello')).not.toBeInTheDocument()" },
+    {
+      code: "expect(rendered.queryAllByText('Hello')).not.toBeInTheDocument()",
+    },
+  ],
+  invalid: [
+    {
+      code: "expect(getByText('Hello')).not.toBeInTheDocument()",
+      errors: [{ messageId: 'expectQueryBy' }],
+    },
+    {
+      code: "expect(rendered.getByText('Hello')).not.toBeInTheDocument()",
+      errors: [{ messageId: 'expectQueryBy' }],
+    },
+    {
+      code: "expect(getAllByText('Hello')).not.toBeInTheDocument()",
+      errors: [{ messageId: 'expectQueryBy' }],
+    },
+    {
+      code: "expect(rendered.getAllByText('Hello')).not.toBeInTheDocument()",
+      errors: [{ messageId: 'expectQueryBy' }],
+    },
+  ],
+});
```

Comment:
You also need to add some tests for autofixing feature, you can find some examples in `tests/lib/rules/no-dom-import.js` and again @thomlom can help you with it if you need something else.

Reply:
Ah didn't know that. I added it now. Thanks

Reply:
Some tests are going to be needed for sure, something like that:

```javascript
    {
      code: `
        const message = getByText('Hello')
        expect(message).toBeInTheDocument() 
      `,
      errors: [{ messageId: 'expectQueryBy' }],
      output: `
        const message = queryByText('Hello')
        expect(message).toBeInTheDocument()
      `,
    },
```

If auto-fixing is complex technically, I think we can leave it as an improvement.

Reply:
@thomlom those specific tests for matched elements saved in a var are not necessary, **but** I found an issue with the current autofix so probably we need to leave it as an improvement.

@emmenko you are fixing the query used inside expect statement, but you also need to take care about if the query is _imported_ or not.

Let's say I have this test:
```
const { getByText } = render(<Foo />);
// ...
expect(getByText('bar')).not.toBeInTheDocument();
```

would be replaced by
```
const { getByText } = render(<Foo />);
// ...
expect(queryByText('bar')).not.toBeInTheDocument();
```
but it's not 100% fixed as `queryByText` is not defined so you would have to destructure/get it from render (which could be harder than it seems as it should be get from destructured object or var declared by the user).

And another case:
```
const { getByText, queryByText } = render(<Foo />);
// ...
expect(getByText('bar')).not.toBeInTheDocument();
```

would be replaced by
```
const { getByText, queryByText } = render(<Foo />);
// ...
expect(queryByText('bar')).not.toBeInTheDocument();
```
so here `queryByText` is already defined and rule doesn't have to change anything related to this.

So here we have to 1) improve the fix function to take care of the new query being defined or not or 2) ignore fixing the rule for now. I'm happy with option 2 and leave the autofix for future improvement as implement the proper autofix here could take longer than expected.

Reply:
Fixing rules do not need to account for other linting errors. As stated in the [eslint docs](https://eslint.org/docs/developer-guide/working-with-rules#applying-fixes)

> Since all rules are run again after the initial round of fixes is applied, it’s not necessary for a rule to check whether the code style of a fix will cause errors to be reported by another rule

<img width="1176" alt="image" src="https://user-images.githubusercontent.com/1110551/66912106-36bdd980-f012-11e9-8237-b123e1bc04ff.png">

So from my understanding, this is totally fine how the rule autofix is implemented regarding its concerns.

Reply:
But this is not about this fixing is breaking another eslint rule: the point is the test won't even work as the new query used could not exist.

Reply:
Yes and breaking the code can also be seen as a eslint error. Of course the user will need to manually fix it somehow.

However, remember the following things:
- this case will only happen the first time you adopt the rule, as you need to "fix/migrate" your code to adhere to it
- autofixing is optional, you don't need to use it if you don't want to

Reply:
I don't agree with "breaking the code can also be seen as a eslint error". Maybe you have some eslint rules covering that particular case to avoid the code to be broken, but if the code was working and because this rule is trying to fix something now the code is broken, that shouldn't be fine at all, even if the autofixing is optional.

Reply:
Also, it's not necessary something happening first time you adopt the rule, as you'll keep adding more tests where you can make the same mistake so it will autofix it again and break the code.

Reply:
I think this "bug" should be addressed. People often make use of the auto fix feature and I think they don't expect ESLint to break their code. I would be annoyed if ESLint introduced a `queryByText` statement and broke my tests while everything was fine before.

As their docs state, fixing should be as small as possible. Thus, I think we should either ship the autofix feature with the `queryByText` imported in case it's not in the destructured statement or just report the code and don't apply fixes. If we leave at-is, it can potentially harm the developer experience.

Reply:
Agree with @thomlom, so I think I prefer to leave the rule with autofix for now and just report the issue. That being said, @emmenko I really appreciate your effort here and we will recover your code for sure when implementing this autofix.

Reply:
Those are all valid arguments and concerns. We can also argue where the line is but it's all very subtle and at the end it comes down to personal preference and expectations.

So the point I'm trying to make is: let's not worry much about it and maybe make it clear (in the docs) what consequences using autofix might lead to. At least it's fair and transparent to the users.
I would also like to keep the autofix implementation relative simple and avoid such complex edge case scenarios.

My experience so far with using this rule, but also in eslint rules in general, is that autofix is only useful the first time I'm trying to adopt a rule. Once the rule is in place, as you're writing code the rule tells your right away if it correct or not, thus the "fixing" part is implicitly part of your current work.

What do you think?

Reply:
Sorry, I understand what you mean, but even if it's explained in the rule info I don't want to leave this autofix with the possibility of breaking the code as it would harm the developer experience.

Reply:
I totally get what you mean and I agree on the fact that fixing rules are only useful the first time.

But this plugin is new and the testing library's ecosystem is still growing. There are some best practices emerging but I think it's still not clear for newcomers to get how to make the most of Testing Library. My concern is that people will rely on this plugin to implement best practices for them automatically, possibly break their tests and have a bad experience with it.
Even if we state it in the docs, this rule will be  in the recommended ones. And I'm pretty sure there will be people who will:

1. Install the plugin (and don't care of the docs since they just want the recommended ones)
2. Run `eslint --fix`
3. See their tests break
4. Remove the plugin
5. Checkout the code

I prefer to have a limited experience that leaves you the manual work instead of a full experience that might cause your code to break.

Reply:
So you want to remove/disable the autofix for now? It's fine by me, I just wanted to share my perspective on this topic before we decide on how to proceed.

Reply:
Of course! This is not about what I want or something like that :) really happy to have your point of view here. Better to remove the fixing implementation (including the badge and the output in the tests, sorry about the last one as I asked that), and think about it in a future issue. For first rules we added to the plugin we also skipped the autofix for most of them as there are some cases we didn't know what to do or it took longer than expected to be implemented, so it's better to go bit by bit.

Reply:
It's disabled here: https://github.com/Belco90/eslint-plugin-testing-library/pull/22/commits/0fd4a6c3e6c968893028c4ba8eee938f45008ffb

I kept it commented out, so that we have a starting point in case we want to pick it up in the future.